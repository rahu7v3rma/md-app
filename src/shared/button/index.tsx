import React, { FunctionComponent } from 'react';
import {
    GestureResponderEvent,
    StyleProp,
    StyleSheet,
    TouchableOpacity,
    ViewStyle
} from 'react-native';

import { Colors } from '@/theme/colors';
import { COMMON } from '@/utils/common';

type Props = {
    children?: React.ReactNode;
    primary?: boolean;
    bordered?: boolean;
    rounded?: boolean;
    block?: boolean;
    style?: StyleProp<ViewStyle>;
    disabled?: boolean;
    onPress?: (e: GestureResponderEvent) => void;
    testID?: string;
};

const Button: FunctionComponent<Props> = ({
    children,
    primary = false,
    bordered = true,
    rounded,
    block,
    style,
    disabled,
    onPress,
    testID
}: Props) => {
    return (
        <TouchableOpacity
            testID={testID}
            style={[
                buttonStyle.wrapper,
                bordered && buttonStyle.bordered,
                primary && buttonStyle.primary,
                disabled && buttonStyle.disabled,
                rounded && buttonStyle.rounded,
                block && buttonStyle.block,
                style
            ]}
            disabled={disabled}
            activeOpacity={disabled ? 1 : 0.75}
            onPress={onPress}
        >
            {children}
        </TouchableOpacity>
    );
};

export default Button;

const buttonStyle = StyleSheet.create({
    wrapper: {
        justifyContent: 'center',
        alignItems: 'center',
        paddingHorizontal: 15.5,
        paddingVertical: COMMON.isIos ? 24 : 17.5,
        borderRadius: 20,
        margin: 0
    },
    bordered: {
        borderWidth: 1,
        borderBottomWidth: 5,
        borderColor: Colors.theme.primary
    },
    primary: {
        backgroundColor: Colors.theme.primary,
        borderColor: Colors.theme.primary
    },
    disabled: {
        backgroundColor: Colors.button.app_button_disabled_bg,
        borderColor: Colors.button.app_button_disabled_border
    },
    rounded: {
        borderRadius: 100,
        backgroundColor: Colors.button.app_btn_light_color
    },
    block: {
        width: '100%'
    }
});
