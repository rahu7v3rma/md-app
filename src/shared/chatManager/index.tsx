import React, { FunctionComponent, useState } from 'react';

import { AppChatProvider } from '@/contexts/appChat';

import { useChatNotifications, useInitChatClient } from './hooks';

type Props = {
    children?: React.ReactNode;
};

const ChatManager: FunctionComponent<Props> = ({ children }: Props) => {
    const { chatClient } = useInitChatClient();

    const [activeChatChannel, setActiveChatChannel] = useState<string | null>(
        null
    );

    useChatNotifications(chatClient, activeChatChannel);

    return (
        <AppChatProvider value={{ chatClient, setActiveChatChannel }}>
            {children}
        </AppChatProvider>
    );
};

export default ChatManager;
