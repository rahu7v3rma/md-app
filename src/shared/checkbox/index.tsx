import React, { FunctionComponent } from 'react';
import { StyleSheet, TouchableOpacity } from 'react-native';

import { CheckIcon } from '@/assets/svgs';
import { Colors } from '@/theme/colors';

type Props = {
    checked: boolean;
    onChange: (value: boolean) => void;
};

const Checkbox: FunctionComponent<Props> = ({ checked, onChange }: Props) => {
    return (
        <TouchableOpacity
            style={[checkboxStyle.wrapper, checked && checkboxStyle.checked]}
            activeOpacity={0.8}
            onPress={() => onChange(!checked)}
        >
            {checked && (
                <CheckIcon color={Colors.extras.white} width={18} height={18} />
            )}
        </TouchableOpacity>
    );
};

export default Checkbox;

const checkboxStyle = StyleSheet.create({
    wrapper: {
        backgroundColor: Colors.text.white,
        borderWidth: 1.5,
        borderColor: Colors.button.app_button_disabled_border,
        width: 28,
        height: 28,
        borderRadius: 9,
        alignItems: 'center',
        justifyContent: 'center'
    },
    checked: {
        backgroundColor: Colors.text.green,
        borderWidth: 0
    }
});
