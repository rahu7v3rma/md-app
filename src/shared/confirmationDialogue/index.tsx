import React, { FunctionComponent, RefObject } from 'react';
import { StyleProp, StyleSheet, View, ViewStyle } from 'react-native';
import RBSheet from 'react-native-raw-bottom-sheet';

import Button from '@/shared/button';
import Text, { Size } from '@/shared/text';
import { Colors } from '@/theme/colors';

type Props = {
    bottomSheetRef: RefObject<RBSheet>;
    title: string;
    subTitle?: string;
    dismissBtnTitle?: string;
    confirmBtnTitle?: string;
    onDismissBtnHandler?: () => void;
    onConfirmBtnHandler?: () => void;
    confirmBtnStyles?: StyleProp<ViewStyle>;
};

const ConfirmationDialogue: FunctionComponent<Props> = ({
    bottomSheetRef,
    title,
    subTitle,
    dismissBtnTitle,
    confirmBtnTitle,
    onDismissBtnHandler,
    onConfirmBtnHandler,
    confirmBtnStyles
}: Props) => {
    return (
        <RBSheet
            ref={bottomSheetRef}
            closeOnDragDown={true}
            animationType={'fade'}
            closeOnPressMask={true}
            customStyles={{
                wrapper: styles.wrapper,
                container: styles.contentContainer,
                draggableIcon: styles.draggableIcon
            }}
        >
            <View style={styles.titlesView}>
                <Text
                    size={Size.Large}
                    fontWeight="bold"
                    color={Colors.text.black_gray}
                    style={styles.title}
                >
                    {title}
                </Text>
                {subTitle && (
                    <Text
                        testID="confirmationDialogueSubTitle"
                        size={Size.XXSmall}
                        color={Colors.text.black}
                        style={styles.subTitle}
                    >
                        {subTitle}
                    </Text>
                )}
            </View>
            <View style={styles.buttonsView}>
                <Button
                    block={true}
                    bordered={false}
                    style={styles.firstButton}
                    onPress={onDismissBtnHandler}
                    testID="dismissBtn"
                >
                    <Text color={Colors.text.primary} fontWeight="600">
                        {dismissBtnTitle}
                    </Text>
                </Button>
                <Button
                    block={true}
                    bordered={false}
                    style={[styles.secondButton, confirmBtnStyles]}
                    onPress={onConfirmBtnHandler}
                    testID="confirmBtn"
                >
                    <Text color={Colors.text.white} fontWeight="600">
                        {confirmBtnTitle}
                    </Text>
                </Button>
            </View>
        </RBSheet>
    );
};

const styles = StyleSheet.create({
    wrapper: {
        backgroundColor: Colors.theme.app_sheet_background_color
    },
    contentContainer: {
        height: 'auto',
        opacity: 1,
        backgroundColor: Colors.extras.white,
        alignItems: 'center',
        borderTopLeftRadius: 40,
        borderTopRightRadius: 40
    },
    draggableIcon: {
        width: 0
    },
    titlesView: {
        width: '80%',
        marginTop: 20
    },
    title: {
        textAlign: 'center'
    },
    subTitle: {
        paddingTop: 20,
        textAlign: 'center'
    },
    buttonsView: {
        width: '100%',
        paddingHorizontal: 20,
        marginTop: 41.86,
        marginBottom: 48
    },
    firstButton: {
        height: 56,
        paddingVertical: 0,
        backgroundColor: Colors.theme.primary_light,
        marginBottom: 16.5
    },
    secondButton: {
        height: 56,
        paddingVertical: 0,
        backgroundColor: Colors.theme.primary,
        borderColor: Colors.button.app_button_primary_border
    }
});

export default ConfirmationDialogue;
