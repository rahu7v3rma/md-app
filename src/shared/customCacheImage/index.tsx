import React, {
    FunctionComponent,
    useCallback,
    useEffect,
    useState
} from 'react';
import { Image, ImageProps, ImageURISource } from 'react-native';
import RNFS from 'react-native-fs';
import { sha256 } from 'react-native-sha256';

type Props = ImageProps & {
    source: ImageURISource;
};

const getLocalUri = (remoteUri: string) => {
    // remove query string from the remote uri when creating the cached path so
    // that all images with the same path are considered the same
    let modifiedRemoteUri = remoteUri.substring(0, remoteUri.lastIndexOf('?'));
    return sha256(modifiedRemoteUri).then(
        (hashedRemoteUri) => `${RNFS.CachesDirectoryPath}/${hashedRemoteUri}`
    );
};

/**
 * This image component should *only* be used for backend-stored content
 * images, which are returned as signed S3 urls that are different every time
 * due to expiration times and url signatures.
 * The component implements a file system cache which ignores the querystring
 * part of urls when caching.
 */
const CustomCacheImage: FunctionComponent<Props> = ({
    source,
    ...props
}: Props) => {
    const [displayUri, setDisplayUri] = useState<string | undefined>(undefined);
    const [fallback, setFallback] = useState<boolean>(false);

    const handleImageChanged = useCallback(
        (localUri: string, remoteUri: string) => {
            RNFS.exists(localUri).then((exists: boolean) => {
                if (exists) {
                    setDisplayUri(`file://${localUri}`);
                } else {
                    const { promise } = RNFS.downloadFile({
                        fromUrl: remoteUri,
                        toFile: localUri
                    });

                    promise
                        .then((result) => {
                            if (result.statusCode === 200) {
                                setFallback(false);
                                setDisplayUri(`file://${localUri}`);
                            } else {
                                console.debug(
                                    'failed to load custom cached image with status',
                                    result.statusCode
                                );
                                setFallback(true);

                                // remove file to make sure we didn't cache any
                                // bad image, ignore errors
                                RNFS.unlink(localUri).catch(() => {});
                            }
                        })
                        .catch((err) => {
                            console.debug(
                                'failed to load custom cached image with error',
                                err
                            );
                            setFallback(true);

                            // remove file to make sure we didn't cache any bad
                            // image, ignore errors
                            RNFS.unlink(localUri).catch(() => {});
                        });
                }
            });
        },
        []
    );

    useEffect(() => {
        const remoteUri = source.uri;

        if (remoteUri) {
            getLocalUri(remoteUri).then((localUri) => {
                handleImageChanged(localUri, remoteUri);
            });
        } else {
            setFallback(true);
        }
    }, [source.uri, handleImageChanged]);

    return (
        <Image
            {...props}
            source={
                fallback ? source : { ...(source as object), uri: displayUri }
            }
        />
    );
};

export default CustomCacheImage;
