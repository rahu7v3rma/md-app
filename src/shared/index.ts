import Button from './button';
import ChatAvatar from './chatAvatar';
import ChatListItem from './chatListItem';
import ChatManager from './chatManager';
import Checkbox from './checkbox';
import ConfirmationDialogue from './confirmationDialogue';
import CustomCacheImage from './customCacheImage';
import CustomStatusBar from './customStatusBar';
import Header from './header';
import ImageChooser from './imageChooser';
import Input from './input';
import LessonCard from './lessonCard';
import ListItem from './listItem';
import LogBookListItem from './logBookListItem';
import LogInputDropdown from './logInputDropdown';
import LogTab from './logTab';
import LogTimePicker from './logTimePicker';
import LogUnitPicker from './logUnitPicker';
import MainTabBar from './mainTabBar';
import NoDataAvailable from './noDataAvailable';
import PdfViewer from './pdfViewer';
import PlatformImage from './platformImage';
import ProfileImage from './profileImage';
import Progressbar from './progressBar';
import ProgressCircle from './progressCircle';
import Switch from './switch';
import Text from './text';
import VideoPlayer from './videoPlayer';
import Wrapper from './wrapper';

export {
    Button,
    ChatAvatar,
    ChatListItem,
    ChatManager,
    Checkbox,
    ConfirmationDialogue,
    CustomCacheImage,
    CustomStatusBar,
    Header,
    ImageChooser,
    Input,
    LessonCard,
    ListItem,
    LogBookListItem,
    LogInputDropdown,
    LogTab,
    LogTimePicker,
    LogUnitPicker,
    MainTabBar,
    NoDataAvailable,
    PlatformImage,
    ProfileImage,
    Progressbar,
    ProgressCircle,
    Switch,
    Text,
    VideoPlayer,
    Wrapper,
    PdfViewer
};
