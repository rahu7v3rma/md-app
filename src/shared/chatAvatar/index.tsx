import React, { FunctionComponent } from 'react';
import { Image, StyleProp, StyleSheet, View, ViewStyle } from 'react-native';

import DefaultAvatar from '@/assets/images/default_avatar.jpg';
import { PlatformImage } from '@/shared';
import { Colors } from '@/theme/colors';

type Props = {
    width?: number;
    height?: number;
    path?: string;
    online?: boolean;
    style?: StyleProp<ViewStyle>;
};

const ChatAvatar: FunctionComponent<Props> = ({
    width,
    height,
    path,
    online,
    style
}: Props) => {
    const borderRadius = width ? width / 2 : 0;

    return (
        <>
            <View style={[styles.wrapper, style]}>
                {path ? (
                    <PlatformImage
                        style={{ borderRadius, width, height }}
                        imageId={path}
                        width={width}
                        height={height}
                    />
                ) : (
                    <Image
                        style={{ borderRadius, width, height }}
                        source={DefaultAvatar}
                    />
                )}
                {online && <View testID="online" style={styles.onlineDot} />}
            </View>
        </>
    );
};

export default ChatAvatar;

const styles = StyleSheet.create({
    wrapper: {
        justifyContent: 'center'
    },
    onlineDot: {
        backgroundColor: Colors.icons.green_check_mark,
        borderStyle: 'solid',
        borderColor: Colors.extras.white,
        width: 12,
        borderWidth: 2,
        height: 12,
        borderRadius: 12,
        zIndex: 1,
        position: 'absolute',
        top: 0,
        transform: [{ translateY: -3 }, { translateX: -3 }]
    }
});
