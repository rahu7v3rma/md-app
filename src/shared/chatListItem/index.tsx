import React, { FunctionComponent } from 'react';
import { StyleSheet, TouchableOpacity, View } from 'react-native';

import { ArrowNext } from '@/assets/svgs';
import { ChatAvatar, Text } from '@/shared';
import { Size } from '@/shared/text';
import { Colors } from '@/theme/colors';
import { LayoutContants } from '@/theme/layoutContants';

type Props = {
    avatar?: any;
    label: string;
    lastMessage?: string;
    time?: string;
    newMessagesCount?: number;
    isOnline?: boolean;
    onPress?: () => void;
    style?: any;
    mainTouchID?: string;
    msgCountID?: string;
};

const ChatListItem: FunctionComponent<Props> = ({
    avatar,
    isOnline = false,
    label,
    lastMessage = '',
    time = '',
    newMessagesCount = 0,
    onPress,
    style,
    mainTouchID,
    msgCountID
}: Props) => {
    return (
        <TouchableOpacity
            testID={mainTouchID}
            onPress={onPress}
            style={[styles.wrapper, style]}
        >
            <View style={styles.firstCol}>
                <View style={styles.avatarView}>
                    <ChatAvatar
                        path={avatar}
                        width={24}
                        height={24}
                        online={isOnline}
                        style={styles.avatar}
                    />
                </View>
            </View>
            <View style={styles.secondCol}>
                <Text
                    size={Size.XSmall}
                    fontWeight="700"
                    color={Colors.text.black}
                >
                    {label}
                </Text>
                {lastMessage !== '' && (
                    <Text size={Size.XXXSmall} color={Colors.text.gray}>
                        {lastMessage.substring(0, 29).concat('...')}
                    </Text>
                )}
            </View>
            <View style={styles.thirdCol}>
                {!time ? (
                    <ArrowNext
                        width={32}
                        height={32}
                        color={Colors.text.purple}
                    />
                ) : (
                    <>
                        <Text size={Size.XXXSmall} color={Colors.text.gray}>
                            {time}
                        </Text>
                        {newMessagesCount > 0 && (
                            <View
                                testID={msgCountID}
                                style={styles.messagesCountView}
                            >
                                <Text color={Colors.text.white}>3</Text>
                            </View>
                        )}
                    </>
                )}
            </View>
        </TouchableOpacity>
    );
};

export default ChatListItem;

const styles = StyleSheet.create({
    wrapper: {
        flexDirection: 'row',
        alignItems: 'center',
        backgroundColor: Colors.extras.white,
        borderRadius: 20,
        borderColor: Colors.button.app_button_disabled_bg,
        height: LayoutContants.cardHeight,
        elevation: 1,
        shadowOpacity: 0.06,
        shadowRadius: 3,
        shadowOffset: { width: 0, height: 0 },
        shadowColor: Colors.theme.app_sheet_background_color
    },
    firstCol: {
        width: '20%',
        alignItems: 'center'
    },
    avatarView: {
        position: 'relative',
        width: 32,
        height: 32,
        borderRadius: 32 / 2
    },
    avatar: {
        width: 32,
        height: 32
    },
    online: {
        position: 'absolute'
    },
    secondCol: {
        width: '60%'
    },
    label: {
        color: Colors.text.black
    },
    thirdCol: {
        width: '20%',
        alignItems: 'center'
    },
    messagesCountView: {
        width: 16,
        height: 16,
        borderRadius: 16 / 2,
        backgroundColor: Colors.theme.primary,
        alignItems: 'center',
        marginTop: 3
    }
});
