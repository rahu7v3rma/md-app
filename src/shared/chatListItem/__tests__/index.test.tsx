import React from 'react';
import renderer, { act } from 'react-test-renderer';

import { ChatListItem } from '@/shared';

it('chatListItem snapshot', async () => {
    const tree = renderer.create(<ChatListItem label={'label'} />);
    expect(tree).toMatchSnapshot();
});

it('chatListItem renders without errors', async () => {
    const tree = renderer.create(<ChatListItem label={'label'} />);
    expect(tree.root).toBeTruthy();
});

it('calls onPress callback when pressed', async () => {
    const onPress = jest.fn();
    const tree = renderer.create(
        <ChatListItem
            onPress={onPress}
            label={'label'}
            mainTouchID={'mainTouchID'}
        />
    );
    let mainTouchID = tree.root.findByProps({
        testID: 'mainTouchID'
    }).props;
    await act(() => mainTouchID.onPress());
    expect(onPress).toHaveBeenCalledTimes(1);
});

it('displays online dot when isOnline prop is true', async () => {
    const tree = renderer.create(
        <ChatListItem label={'label'} isOnline={true} />
    );
    let onlineView = tree.root.findByProps({
        testID: 'online'
    }).props;
    expect(onlineView).toBeTruthy();
});

it('displays newMessagesCount when it is greater than 0', async () => {
    const tree = renderer.create(
        <ChatListItem
            label={'label'}
            newMessagesCount={3}
            time={'12:00'}
            msgCountID={'msgCountID'}
        />
    );
    let msgCountID = tree.root.findByProps({
        testID: 'msgCountID'
    }).props;
    expect(msgCountID).toBeTruthy();
});
