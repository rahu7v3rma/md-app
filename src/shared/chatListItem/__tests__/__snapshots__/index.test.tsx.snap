// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chatListItem snapshot 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "borderColor": "#cccccc",
      "borderRadius": 20,
      "elevation": 1,
      "flexDirection": "row",
      "height": 90,
      "opacity": 1,
      "shadowColor": "rgba(0,0,0,0.5)",
      "shadowOffset": {
        "height": 0,
        "width": 0,
      },
      "shadowOpacity": 0.06,
      "shadowRadius": 3,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "width": "20%",
      }
    }
  >
    <View
      style={
        {
          "borderRadius": 16,
          "height": 32,
          "position": "relative",
          "width": 32,
        }
      }
    >
      <View
        style={
          [
            {
              "justifyContent": "center",
            },
            {
              "height": 32,
              "width": 32,
            },
          ]
        }
      >
        <Image
          source={
            {
              "testUri": "../../../src/assets/images/default_avatar.jpg",
            }
          }
          style={
            {
              "borderRadius": 12,
              "height": 24,
              "width": 24,
            }
          }
        />
      </View>
    </View>
  </View>
  <View
    style={
      {
        "width": "60%",
      }
    }
  >
    <Text
      adjustsFontSizeToFit={true}
      style={
        [
          {
            "fontFamily": "Poppins",
          },
          undefined,
          {
            "color": "#4D4F4D",
            "fontSize": 14,
            "fontWeight": "bold",
            "lineHeight": 22,
          },
        ]
      }
    >
      label
    </Text>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "width": "20%",
      }
    }
  >
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight={32}
      bbWidth={32}
      fill="none"
      focusable={false}
      height={32}
      meetOrSlice={0}
      minX={0}
      minY={0}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          {
            "flex": 0,
            "height": 32,
            "width": 32,
          },
        ]
      }
      vbHeight={24}
      vbWidth={24}
      width={32}
    >
      <RNSVGGroup
        fill={null}
        propList={
          [
            "fill",
          ]
        }
      >
        <RNSVGPath
          d="M9.5 7L14.5 12L9.5 17"
          propList={
            [
              "stroke",
              "strokeWidth",
              "strokeLinecap",
              "strokeLinejoin",
            ]
          }
          stroke={4285419984}
          strokeLinecap={1}
          strokeLinejoin={1}
          strokeWidth={2}
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
</View>
`;
